osp/core/ontology/utils.py

```diff
@@ -0,0 +1,284 @@
+"""Utility resources for the ontology module."""
+
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, Iterator, MutableSet, Set
+
+
+class DataStructureSet(ABC, MutableSet):
+    """As set-like object that acts as an interface to another data structure.
+
+    This class looks like and acts like the standard `set`, but the data it
+    holds is stored on a different data structure. When an instance is read
+    or when it is modified in-place, specific methods defined to interact
+    with the data structure are called to reflect the changes.
+
+    This class does not hold any object-related information itself, thus
+    it is safe to spawn multiple instances linked to the same data
+    structure (when single-threading).
+    """
+
+    @abstractmethod
+    def __iter__(self) -> Iterator[Any]:
+        """Implement iter(self).
+
+        Returns the associated values from the connected data structure.
+
+        The values provided MUST NOT be repeated, as instances of this class
+        are expected to behave like sets. Remember that the definition of
+        membership for sets in Python is `any(x is e or x == e for e in y)`,
+        see https://docs.python.org/3/reference/expressions.html
+        #membership-test-details. Therefore, if you implemented this method
+        correctly, letting `y` be this object, then
+        `sum(sum(x is e or x == e for e in y) for x in y)` should evaluate to
+        `len(list(y))`.
```

Comment:
I think pointing the definition of set membership in Python (`any(x is e or x == e for e in y)`) is still relevant, that is, defining what does it mean to be `unique`, because people may check one of the two conditions rather than both. I'll try to rephrase this in terms of the definition of unique.